accord-core/src/main/java/accord/coordinate/Barrier.java

```diff
@@ -0,0 +1,309 @@
+            {
+                Timestamp executeAt = command.executeAt();
+                // In all the cases where we add a listener (listening to existing command, async completion of CoordinateSyncPoint)
+                // we want to notify the agent
+                doBarrierSuccess(executeAt);
+            }
+            else if (command.is(Status.Invalidated))
+                // Surface the invalidation and the barrier can be retried by the caller
+                Barrier.this.tryFailure(new Timeout(command.txnId(), command.homeKey()));
+
+        }
+
+        @Override
+        public PreLoadContext listenerPreLoadContext(TxnId caller)
+        {
+            return PreLoadContext.contextFor(caller);
+        }
+    }
+
+    private ExistingTransactionCheck checkForExistingTransaction()
+    {
+        checkState(seekables.size() == 1 && seekables.domain() == Domain.Key);
+        ExistingTransactionCheck check = new ExistingTransactionCheck();
+        Key k = seekables.get(0).asKey();
+        node.commandStores().mapReduceConsume(
+                contextFor(k),
+                k.toUnseekable(),
+                minEpoch,
+                Long.MAX_VALUE,
+                check);
+        return check;
+    }
+
+    // Hold result of looking for a transaction to act as a barrier for an Epoch
+    static class BarrierTxn
+    {
+        @Nonnull
+        public final TxnId txnId;
+        @Nonnull
+        public final Timestamp executeAt;
+        @Nonnull
+        public final Status status;
+        @Nonnull
+        public final Key key;
+        public BarrierTxn(@Nonnull TxnId txnId, @Nonnull Timestamp executeAt, @Nonnull Status status, Key key)
+        {
+            this.txnId = txnId;
+            this.executeAt = executeAt;
+            this.status = status;
+            this.key = key;
+        }
+
+        public BarrierTxn max(BarrierTxn other)
+        {
+            if (other == null)
+                return this;
+            return status.compareTo(other.status) >= 0 ? this : other;
+        }
+    }
+
+    /*
+     * Check for an existing transaction that is either already Applied (dependencies were applied)
+     * or PreApplied (outcome fixed, dependencies not yet applied).
+     *
+     * For Applied we can return success immediately with the executeAt epoch. For PreApplied we can add
+     * a listener for when it transitions to Applied and then return success.
+     */
+    class ExistingTransactionCheck extends AsyncResults.AbstractResult<BarrierTxn> implements MapReduceConsume<SafeCommandStore, BarrierTxn>
+    {
+        @Override
+        public BarrierTxn apply(SafeCommandStore safeStore)
+        {
+            BarrierTxn found = safeStore.mapReduceWithTerminate(
+                    seekables,
+                    safeStore.ranges().allSince(minEpoch),
+                    // TODO review this with Benedict, but I don't believe an exclusive sync point works here
```

Comment:
is this TODO still relevant?

Reply:
@belliottsmith For a barrier where we want to know that everything from before that transaction has applied already would exclusive sync points work?

I think they would because they wouldn't enter the `Applied` state until their dependencies were applied and they do take the necessary dependencies?

Reply:
Define before? An `ExclusiveSyncPoint` defines an order on `txnId` not `executeAt` - their `executeAt` is sort of imprecise, somewhere after its own `txnId` and its dependency with the highest `executeAt` - logically different on each replica and key.

Reply:
`local` barrier (this path) only need to establish that no transactions that execute in epochs < `minEpoch` are unapplied. It's OK if transactions that execute in the current epoch or later are not applied. `local` barriers will try to use an existing transaction to establish that the constraint is satisfied and if they can't they create a `SyncPoint` transaction.

`global` barriers need to establish a happens before relationship between any observed side effects of transactions from before the creation of the barrier, and the application of those transactions. The two use cases are ensuring no accord transactions from an older epoch are unapplied (migration) and ensuring that any transaction side effects witnessed by repair are from now committed transactions. Global barriers always create a `SyncPoint` transaction and then wait on it.

If a `SyncPoint` and `ExclusiveSyncPoint` both have the property that once they are applied than any transaction with a lower `txnId` has applied then they should both be viable transaction types to wait on instead of creating a new `SyncPoint` transaction.

When searching for a transaction the parameters provided are `TestTimestamp.EXECUTES_AFTER` and `TxnId.minForEpoch(minEpoch)` so I thought that meant I was picking a transaction with an established `executeAt` that once applied also means the application of any transaction with a `txnId` or `executeAt` <  the selected transaction and that it would work with read and read/write transactions.

Reply:
> once applied also means the application of any transaction with a txnId or executeAt < the selected transaction

By definition, txnId < executeAt.

> If a SyncPoint and ExclusiveSyncPoint both have the property that once they are applied than any transaction with a lower txnId has applied

Only an ExclusiveSyncPoint has this property, as a lower `txnId` may not have yet agreed to execute when the `SyncPoint` was agreed, and the standard `SyncPoint` does not invalidate the earlier `txnId`. However, such a `txnId` would execute logically after the `SyncPoint`. We don't make any specific guarantees about `executeAt` because we don't synchronise across the whole range simultaneously, it's a partial order. But I think this satisfies your required properties.